### PR TITLE
PR for the CFPB Consumer Complaints dataset

### PR DESCRIPTION
**Pull request by kayvane1**
Think I followed all the steps but please let me know if anything needs changing or any improvements I can make to the code quality

### CODE REVIEW

File: datasets/consumer-finance-complaints/README.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+- crowdsourced
+languages:
+- en-US
+licenses:
+- upl-1.0
+multilinguality:
+- monolingual
+pretty_name: consumer-finance-complaints
+size_categories:
+- unknown
```
Comment (lhoestq) [inline datasets/consumer-finance-complaints/README.md:14]:
```suggestion
- 1M<n<10M
```

File: datasets/consumer-finance-complaints/README.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+## Additional Information
+
+### Dataset Curators
+
+https://cfpb.github.io/api/ccdb/
+
+### Licensing Information
+
+https://cfpb.github.io/source-code-policy/
```
Comment (lhoestq) [inline datasets/consumer-finance-complaints/README.md:171]: It sounds like the policy for source code of the Consumer Complaint Database API that is open source, is there some lincensing/terms of use for the data themselves ?

File: datasets/consumer-finance-complaints/README.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+  - [Other Known Limitations](#other-known-limitations)
+- [Additional Information](#additional-information)
+  - [Dataset Curators](#dataset-curators)
+  - [Licensing Information](#licensing-information)
+  - [Citation Information](#citation-information)
+  - [Contributions](#contributions)
+
+## Dataset Description
+
+- **Homepage:**
```
Comment (lhoestq) [inline datasets/consumer-finance-complaints/README.md:52]:
```suggestion
- **Homepage:** https://www.consumerfinance.gov/data-research/consumer-complaints/
```

File: datasets/consumer-finance-complaints/README.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+| Submitted via   | How the complaint was submitted to the CFPB        | plain text        | This field is a categorical variable.        |
+| Date sent to company   | The date the CFPB sent the complaint to the company        | date & time        |         |
+| Company response to consumer   | This is how the company responded. For example, "Closed with explanation."        | plain text        | This field is a categorical variable.        |
+| Timely response?   | Whether the company gave a timely response        | plain text        | yes/no        |
+| Consumer disputed?   | Whether the consumer disputed the company’s response        | plain text        | YES/ NO/ N/A: The Bureau discontinued the consumer dispute option on April 24, 2017.        |
+| Complaint ID   | The unique identification number for a complaint        | number        |         |
```
Comment (lhoestq) [inline datasets/consumer-finance-complaints/README.md:80]: Can this table be moved under `Data Fields` ?

File: datasets/consumer-finance-complaints/README.md [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+## Dataset Description
+
+- **Homepage:**
+- **Repository:**
+- **Paper:**
+- **Leaderboard:**
+- **Point of Contact:**
+
+### Dataset Summary
```
Comment (lhoestq) [inline datasets/consumer-finance-complaints/README.md:58]: I think we should have a brief summary in the beginning here:

```suggestion
### Dataset Summary

This database is a collection of complaints about consumer financial products and services that we sent to companies for response.
```

File: datasets/consumer-finance-complaints/consumerComplaints.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    "Consumer Loan",
+    "Payday loan",
+    "Money transfers",
+    "Prepaid card",
+    "Other financial service",
+    "Virtual currency",
+]
+
+
+class consumerComplaints(datasets.GeneratorBasedBuilder):
```
Comment (lhoestq) [inline datasets/consumer-finance-complaints/consumerComplaints.py:144]:
```suggestion
class ConsumerComplaints(datasets.GeneratorBasedBuilder):
```

File: datasets/consumer-finance-complaints/consumerComplaints.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+        return datasets.DatasetInfo(
+            description=_DESCRIPTION,
+            features=features,
+            homepage=_HOMEPAGE,
+            license=_LICENSE,
+            citation=_CITATION,
+            task_templates=[
+                TextClassification(text_column="Complaint Text", label_column="Product"),
+                TextClassification(text_column="Complaint Text", label_column="Sub Product"),
+                TextClassification(text_column="Complaint Text", label_column="Tags"),
```
Comment (lhoestq) [inline datasets/consumer-finance-complaints/consumerComplaints.py:179]: Awesome !

File: datasets/consumer-finance-complaints/consumerComplaints.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+            ],
+        )
+
+    def _split_generators(self, dl_manager):
+        path = dl_manager.download_and_extract(_url)
+        path = f"{path}/complaints.csv"
```
Comment (lhoestq) [inline datasets/consumer-finance-complaints/consumerComplaints.py:185]: Please use `os.path.join` or you might have issues on windows (the separator is a backslash, not a slash) or in streaming mode (we use [URL chaining](https://filesystem-spec.readthedocs.io/en/latest/features.html#url-chaining))
```suggestion
        path = os.path.join(path, "complaints.csv")
```

**Review by lhoestq — COMMENTED** [review]
Nice ! Thanks for adding this dataset :)

I left a few comments:

**Comment by kayvane1** [comment]
> Nice ! Thanks for adding this dataset :)
> 
> 
> 
> I left a few comments:

Thanks!

I'd be interested in contributing to the core codebase - I had to go down the custom loading approach because I couldn't pull this dataset in using the load_dataset() method. Using either the json or csv files available for this dataset as it was erroring. 

I'll rerun it and share the errors and try debug

**Comment by kayvane1** [comment]
Hey @lhoestq ,

When I use this dataset as part of my project, I'm using this method

`text_dataset = text_dataset['train'].train_test_split(test_size=0.2)`

to create a train and test split as this dataset doesn't have one. 

Can I add this directly in the script itself somehow, or is it better to give users the flexibility to slice and split their datasets after loading?

**Comment by lhoestq** [comment]
> I'd be interested in contributing to the core codebase - I had to go down the custom loading approach because I couldn't pull this dataset in using the load_dataset() method. Using either the json or csv files available for this dataset as it was erroring.
>
> I'll rerun it and share the errors and try debug

Cool ! Let me know if you have questions or if I can help :)

> Can I add this directly in the script itself somehow, or is it better to give users the flexibility to slice and split their datasets after loading?

Usually we let the users the flexibility to split the datasets themselves (unless the dataset is already split, or if there is already a standard way to split it in the papers that use it)

File: datasets/consumer-finance-complaints/consumerComplaints.py [hunk 9]
```diff
@@ -0,0 +1,240 @@
+import csv
+import textwrap
```
Comment (lhoestq) [inline datasets/consumer-finance-complaints/consumerComplaints.py:2]: I think an import is missing here:
```suggestion
import os
import textwrap
```

Comment (lhoestq) [reply datasets/consumer-finance-complaints/consumerComplaints.py:2]: Can you also rename this file `consumer-finance-complaints.py` please ? It must have the same name as the directory

**Comment by kayvane1** [comment]
Thanks Quentin!
All okay to merge now?

**Review by mariosasko — COMMENTED** [review]
Nice work! A few nits from me. Also, could you please do the following before we merge:
* generate dummy data (instructions: https://huggingface.co/docs/datasets/master/dataset_script.html#dummy-data)
* regenerate the `dataset_info` file to have a single line of JSON
* pull the changes from our master branch to make the Windows CI green:
   ```bash
   git remote add upstream https://github.com/huggingface/datasets.git
   git pull upstream master
   git push
   ```

**Comment by kayvane1** [comment]
Thanks for the feedback Quentin and Mario - implemented all changes  :)
Screenshot 2022-01-31 at 23 11 20

**Comment by kayvane1** [comment]
Hey @lhoestq  / @mariosasko 
Any other changes required to merge? 🤗

**Comment by lhoestq** [comment]
Hi ! Thanks and sorry for the late response 

It looks very good ! The CI is still failing because it can't file the dummy_data.zip file, you can fix that by moving `datasets/consumer-finance-complaints/dummy/1.0.0/dummy_data.zip` to `datasets/consumer-finance-complaints/dummy/0.0.0/dummy_data.zip` and it should be all good !

**Comment by kayvane1** [comment]
@lhoestq - hopefully that should do it!

File: datasets/consumer-finance-complaints/consumer-finance-complaints.py [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+The Consumer Complaint Database is a collection of complaints about consumer financial products and services that we sent to companies for response. Complaints are published after the company responds, confirming a commercial relationship with the consumer, or after 15 days, whichever comes first. Complaints referred to other regulators, such as complaints about depository institutions with less than $10 billion in assets, are not published in the Consumer Complaint Database. The database generally updates daily.
+There are multiple Text Classification problems that can be solved with this dataset:
+- Complaint Type Identification
+- Complaint Sub-Product Identification
+- Complaint from vulnerable or Service-person
+"""
+)
+
+_CITATION = """\
+@Consumer Financial Protection Bureau
```
Comment (lhoestq) [inline datasets/consumer-finance-complaints/consumer-finance-complaints.py:23]: we can keep it empty
```suggestion
```

File: datasets/consumer-finance-complaints/dataset_infos.json [hunk 11]
```diff
@@ -0,0 +1 @@
...ype": "string", "id": null, "_type": "Value"}, "Timely Response": {"dtype": "string", "id": null, "_type": "Value"}, "Consumer Disputed": {"dtype": "string", "id": null, "_type": "Value"}, "Complaint ID": {"dtype": "string", "id": null, "_type": "Value"}}, "post_processed": null, "supervised_keys": null, "task_templates": [{"task": "text-classification", "text_column": "Complaint Text", "label_column": "Product"}, {"task": "text-classification", "text_column": "Complaint Text", "label_column": "Sub Product"}, {"task": "text-classification", "text_column": "Complaint Text", "label_column": "Tags"}], "builder_name": "consumer_complaints", "config_name": "default", "version": {"version_str": "0.0.0", "description": null, "major": 0, "minor": 0, "patch": 0}, "splits": {"train": {"name": "train", "num_bytes": 1605177353, "num_examples": 2455765, "dataset_name": "consumer_complaints"}}, "download_checksums": {"https://files.consumerfinance.gov/ccdb/complaints.csv.zip": {"num_bytes": 404187716, "checksum": "19a047bf33e413528b2dcb7780020eb4b0739ec13ea95ccc3f638d158a4ef6d3"}}, "download_size": 404187716, "post_processing_size": null, "dataset_size": 1605177353, "size_in_bytes": 2009365069}}
```
Comment (lhoestq) [inline datasets/consumer-finance-complaints/dataset_infos.json:1]: set citation to none
```suggestion
{"default": {"description": "The Consumer Complaint Database is a collection of complaints about consumer financial products and services that we sent to companies for response. Complaints are published after the company responds, confirming a commercial relationship with the consumer, or after 15 days, whichever comes first. Complaints referred to other regulators, such as complaints about depository institutions with less than $10 billion in assets, are not published in the Consumer Complaint Database. The database generally updates daily.\nThere are multiple Text Classification problems that can be solved with this dataset:\n- Complaint Type Identification\n- Complaint Sub-Product Identification\n- Complaint from vulnerable or Service-person\n", "citation": "\n", "homepage": "https://www.consumerfinance.gov/data-research/consumer-complaints/", "license": "https://cfpb.github.io/source-code-policy/", "features": {"Date Received": {"dtype": "timestamp[s]", "id": null, "_type": "Value"}, "Product": {"num_classes": 18, "names": ["Credit reporting, credit repair services, or other personal consumer reports", "Debt collection", "Mortgage", "Credit card or prepaid card", "Checking or savings account", "Credit reporting", "Student loan", "Money transfer, virtual currency, or money service", "Credit card", "Vehicle loan or lease", "Bank account or service", "Payday loan, title loan, or personal loan", "Consumer Loan", "Payday loan", "Money transfers", "Prepaid card", "Other financial service", "Virtual currency"], "names_file": null, "id": null, "_type": "ClassLabel"}, "Sub Product": {"num_classes": 77, "names": ["Credit reporting", "General-purpose credit card or charge card", "Checking account", "Other debt", "Second mortgage", "Conventional home mortgage", "I do not know", "Credit card debt", "Medical debt", "Federal student loan servicing", "FHA mortgage", "Conventional fixed mortgage", "Loan", "Other (i.e. phone, health club, etc.)", "Store credit card", "Installment loan", "Credit card", "Medical", "Mobile or digital wallet", "Private student loan", "Non-federal student loan", "Domestic (US) money transfer", "VA mortgage", "Vehicle loan", "Auto debt", "Payday loan", "Conventional adjustable mortgage (ARM)", "Other personal consumer report", "Payday loan debt", "Savings account", "Virtual currency", "Other bank product/service", "Other type of mortgage", "Other banking product or service", "Other mortgage", "International money transfer", "Lease", "General-purpose prepaid card", "Home equity loan or line of credit (HELOC)", "Government benefit card", "Mortgage debt", "Personal line of credit", "Home equity loan or line of credit", "Federal student loan debt", "Private student loan debt", "Credit repair services", "Title loan", "Auto", "Vehicle lease", "Mortgage", "Reverse mortgage", "General purpose card", "CD (Certificate of Deposit)", "Federal student loan", "Payroll card", "Debt settlement", "Check cashing service", "Traveler's check or cashier's check", "Gift card", "(CD) Certificate of deposit", "Money order", "Foreign currency exchange", "Refund anticipation check", "Gift or merchant card", "Cashing a check without an account", "ID prepaid card", "Mobile wallet", "Government benefit payment card", "Pawn loan", "Other special purpose card", "Check cashing", "Credit repair", "Traveler\u2019s/Cashier\u2019s checks", "Transit card", "Student prepaid card", "Electronic Benefit Transfer / EBT card", ""], "names_file": null, "id": null, "_type": "ClassLabel"}, "Issue": {"dtype": "string", "id": null, "_type": "Value"}, "Sub Issue": {"dtype": "string", "id": null, "_type": "Value"}, "Complaint Text": {"dtype": "string", "id": null, "_type": "Value"}, "Company Public Response": {"dtype": "string", "id": null, "_type": "Value"}, "Company": {"dtype": "string", "id": null, "_type": "Value"}, "State": {"dtype": "string", "id": null, "_type": "Value"}, "Zip Code": {"dtype": "string", "id": null, "_type": "Value"}, "Tags": {"num_classes": 4, "names": ["Servicemember", "Older American", "Older American, Servicemember", ""], "names_file": null, "id": null, "_type": "ClassLabel"}, "Consumer Consent Provided": {"dtype": "string", "id": null, "_type": "Value"}, "Submitted via": {"dtype": "string", "id": null, "_type": "Value"}, "Date Sent To Company": {"dtype": "string", "id": null, "_type": "Value"}, "Company Response To Consumer": {"dtype": "string", "id": null, "_type": "Value"}, "Timely Response": {"dtype": "string", "id": null, "_type": "Value"}, "Consumer Disputed": {"dtype": "string", "id": null, "_type": "Value"}, "Complaint ID": {"dtype": "string", "id": null, "_type": "Value"}}, "post_processed": null, "supervised_keys": null, "task_templates": [{"task": "text-classification", "text_column": "Complaint Text", "label_column": "Product"}, {"task": "text-classification", "text_column": "Complaint Text", "label_column": "Sub Product"}, {"task": "text-classification", "text_column": "Complaint Text", "label_column": "Tags"}], "builder_name": "consumer_complaints", "config_name": "default", "version": {"version_str": "0.0.0", "description": null, "major": 0, "minor": 0, "patch": 0}, "splits": {"train": {"name": "train", "num_bytes": 1605177353, "num_examples": 2455765, "dataset_name": "consumer_complaints"}}, "download_checksums": {"https://files.consumerfinance.gov/ccdb/complaints.csv.zip": {"num_bytes": 404187716, "checksum": "19a047bf33e413528b2dcb7780020eb4b0739ec13ea95ccc3f638d158a4ef6d3"}}, "download_size": 404187716, "post_processing_size": null, "dataset_size": 1605177353, "size_in_bytes": 2009365069}}
```

File: datasets/consumer-finance-complaints/README.md [hunk 12]
```diff
@@ -0,0 +1,214 @@
+---
+annotations_creators:
+- crowdsourced
+language_creators:
+- crowdsourced
+languages:
+- en-US
```
Comment (lhoestq) [inline datasets/consumer-finance-complaints/README.md:7]:
```suggestion
- en
```

**Review by lhoestq — APPROVED** [review]
Perfect thanks !

**Pull request merged**